README.md
# Python-Validate
python表单验证,类似于jQuery validation前端验证，其中ipv4和ipv6的验证参考了jQuery validation中的验证。通常用于后台开发验证前端的数据。目前支持 `python3`,`python2可以正常使用，但是返回的提示是unicode编码.`

# 1. 安装

## 1.1 通过 pip 安装
```sh
pip install python-validate
```


## 1.2 通过whl安装
下载地址： 
[python_validate-0.0.2.dev0-py2.py3-none-any.whl](https://pypi.python.org/packages/c9/5a/0163dfbd42bda5af13159d3b1c1d865faf0b41359eea06962e3523cfa039/python_validate-0.0.2.dev0-py2.py3-none-any.whl)

# 2.使用
```python
from validate import check

# 验证身份证号码
check.id_card('522130199310062011')
# True

check.id_card('522130199310062012')
# False

check.ipv4('192.168.1.1')
# True
```

# 2.1.常用方法列表：
```
check.id_card           #字符串，验证是否是中国公民身份证号码,check.id_card('522130199310062012') # False
check.enum              #第一个是任何类型a，第二个是集合b，验证a是否在b中，check.enum('1',[1,2,3]) #False
check.url               #字符串，验证是否是合法的url,check.url('http://wwwbaidu.com') #True
check.email             #字符串，验证电子邮件地址，check.email('rean.ren@qq.com') #True
check.equal_to          #比较两个参数是否相等，check.equal_to('a','b') #False

check.min_value         #验证a是否小于最小值b，先转换为数值比较，如果转换不成功，则转换为字符串比较,check.min_value(1,3) #False
chech.max_value         #验证a是否大于最大值b,先转换为数值比较，如果转换不成功，则转换为字符串比较,check.max_value(1,3) #True
check.range_value       #验证a是否在范围b内（元组或者数组）,先转换为数值比较，如果转换不成功，则转换为字符串比较,check.rang_value(2,[1,3]) #True

check.min_length        #验证a的长度是否小于给定值b,check.min_length('22',1) #True
check.max_length        #验证a的长度是否大于给定值b,check.max_length('22',1) #False
check.range_length      #验证a的长度是否在范围b内（元组或者数组）,check.range_length('22',(1,3)) #True

check.pattern           #验证b是否符合正则表达式a,check.pattern('www','www.baidu.com') #False
check.ipv4              #验证ipv4地址，check.ipv4('192.168.1.1') #True
check.ipv6              #验证ipv6地址，check.ipv6('2001:0:9d38:953c:2486:2115:3f57:5f07') #True
check.tel               #验证11位中国电话号码， check.tel('13984927617') #True
check.post_code         #验证6位中国邮编，check.post_code('550025') #True
```

# 2.2 时间格式验证
```
check.date              #验证时间格式，第一个参数a为要验证的字符串，第二个为要验证规则，举例如下：
a='2018-01-04'
birth = {
        'format': '%Y-%m-%d', 
        'min': 1,  # 1，最小值是当前时间，其他字符串
        'max': '2018-01-31',  # 1，最大值是当前时间，其他字符串
        'range': ['2018-01-01', '2018-01-31']
}
check.date(a,birth) #False min没有通过
```

* format 为时间格式，如下：
```
%y 两位数的年份表示（00-99）
%Y 四位数的年份表示（000-9999）
%m 月份（01-12）
%d 月内中的一天（0-31）
%H 24小时制小时数（0-23）
%I 12小时制小时数（01-12）
%M 分钟数（00=59）
%S 秒（00-59）
%a 本地简化星期名称
%A 本地完整星期名称
%b 本地简化的月份名称
%B 本地完整的月份名称
%c 本地相应的日期表示和时间表示
%j 年内的一天（001-366）
%p 本地A.M.或P.M.的等价符
%U 一年中的星期数（00-53）星期天为星期的开始
%w 星期（0-6），星期天为星期的开始
%W 一年中的星期数（00-53）星期一为星期的开始
%x 本地相应的日期表示
%X 本地相应的时间表示
%Z 当前时区的名称
%% %号本身
```
* min 时间的最小值，如果是1则代表当前时间
* max 时间最大值，如果是1则代表当前时间
* range 时间最小值最大值的范围
* min,max,range中的时间格式要和format格式一样

# 3.数据验证

使用check.validate(a,b)进行数据验证,为验证的规则，b为验证的对象。参见[example.py](example.py).

# 3.1 规则
规则字典包含三个键：
* all_validate              True或者False,如果是True，表示验证全部字段，False,只要有一个字段不通过就返回。
* fields                    dict，键为验证的字段，值为验证的规则，规则如下
```
required            True或者False,True表示必填，如果要验证该属性，则需要将其放在第一位，如果没有该属性但是有其他的属性，则会默认验证该属性
enum                使用enum函数验证
type                验证数据类型，常用的有int和str
email               使用email函数验证
url                 使用url函数验证
equal_to            使用equal_to函数验证
min                 使用min_value函数验证
max                 使用max_value函数验证
range               使用range_value函数验证
min_length          使用min_length函数验证
max_length          使用max_length函数验证
range_length        使用range_length函数验证
pattern             使用pattern函数验证
custom              自定义函数验证，函数参数为field(字段名),value(字段值),如果验证成功返回True,否则返回False
id_card             使用id_card函数验证
ipv4                True或者False,True使用ipv4函数验证
ipv6                True或者False,True使用ipv6函数验证
date                使用date函数验证
tel                 True或者False,True使用tel函数验证
post_code           True或者False,True使用post_cdoe函数验证
``` 
* msg dict,验证不通过的提示信息，键和fields一致

如下验证：
```
def custom(filed, value):
    """
    自定义验证函数
    :param filed:字段名
    :param value:字段值
    :return:
    """
    if filed == 'ipaddr':
        return check.pattern(value, '^(25[0-5]|2[0-4]\d|[0-1]?\d?\d)(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}$')


rule = {
        'all_validate': True,  # 全部验证在返回，否则只要有错就返回
        'fields': {
            'name': {
                'required': True,
                'enum': ['rean', 'huen'],
                'type': str,
                'range': [1, 100],
                'range_length': [2, 10],
            },
            'password': {
                'email': True,
                'range_length': [16, 20]
            },
            'confirm_password': {
                'equal_to': 'password',
            },
            'home_page': {
                'url': True
            },
            'score': {
                'range': [200, 2000]
            },
            'address': {
                'min_length': 4,
                'max_length': 10
            },
            'address1': {
                'range_length': [2, 10]
            },
            'ipaddr': {
                'custom': custom  #自定义函数验证
            },
            'birth': {
                'date': {
                    'format': '%Y-%m-%d',
                    'min': 1,  # 1，最小值是当前时间，其他字符串
                    'max': '2018-01-31',  # 1，最大值是当前时间，其他字符串
                    'range': ['2018-01-01', '2018-01-31']
                }
            }
        },
        'msg': {
            'name': {
                'enum': '姓名必须在huen和rean中'
            },
            'confirm_password': {
                'equal_to': '两次输入的密码不一样。'
            }
        }
    }

```


## 3.2 提示信息

* 设置全局提示信息：
```
#覆盖默认提示
check.default_msg={
    'required':'字段必填'
}
```
* 设置字段指定规则的提示
```
rule={
    'msg':{
        'name':{
            'rquired':'请填写姓名'
    }
}
```
* 设置字段默认提示
```
rule={
    'msg':{
        'name':{
            '_default':'姓名填写不正确',
            'rquired':'请填写姓名'
    }
}
```

* 消息查找顺序
    * 1.自定义每个属性的消息
    * 2.字段默认消息（_default）
    * 3.全局默认消息（check.default_msg）
# 3.3 返回参数
返回参数是一个元组，三个值。
* 第一个是True或者False,表示验证成功或者失败
* 第二个是一个list，验证失败的字段,每一个元素是一个元组，分别是字段名和提示信息
* 第三个是一个list,验证成功的字段
```
(False, [('name', '默认消息提示'), ('password', '字段的长度不在范围内.'), ('score', '字段的值不在范围内.'), ('address', '字段太长.'), ('address1', '字段的长度不在范围内.'), ('birth', '生日输入不正确.')], ['confirm_password', 'home_page', 'ipaddr'])
```


# 3.4 完整例子

请参考[example.py](example.py).

# 4.欢迎提issue和PR
有更好的验证方法和建议，欢迎提issue和PR，welcome!


ChangeLog.txt
# python-validate changelog

## 2018年1月5日 0.0.2-dev
first public version.

test.py
from validate import check


def custom(filed, value):
    """
    自定义验证函数
    :param filed:字段名
    :param value:字段值
    :return:
    """
    if filed == 'ipaddr':
        return check.pattern(value, '^(25[0-5]|2[0-4]\d|[0-1]?\d?\d)(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}$')


if __name__ == '__main__':
    # 覆盖默认提示
    check.default_msg['required'] = '字段必填'
    msg = {
        'name': {
            # 默认消息提示，如果找不到下一级的消息提示则提示该消息
            '_default': '默认消息提示',

            # 覆盖默认提示
            'enum': '姓名必须在huen和rean中'
        },
        'confirm_password': {
            'equal_to': '两次输入的密码不一样。'
        }
    }
    # 消息查找顺序
    # 1.自定义每个属性的消息
    # 2.字段默认消息（_default）
    # 3.全局默认消息

    rule = {
        'all_validate': True,  # 全部验证在返回，否则只要有错就返回
        'fields': {
            'name': {
                'required': True,
                'enum': ['rean', 'huen'],
                'type': str,
                'range': [1, 100],
                'range_length': [2, 10],
            },
            'password': {
                'email': True,
                'range_length': [16, 20]
            },
            'confirm_password': {
                'equal_to': 'password',
                'message': '两次的密码不一样'
            },
            'home_page': {
                'url': True
            },
            'score': {
                'range': [200, 2000]
            },
            'address': {
                'min_length': 4,
                'max_length': 10
            },
            'address1': {
                'range_length': [2, 10]
            },
            'ipaddr': {
                'custom': custom
            },
            'birth': {
                'date': {
                    'format': '%Y-%m-%d',
                    'min': 1,  # 1，最小值是当前时间，其他字符串
                    'max': '2018-01-31',  # 1，最大值是当前时间，其他字符串
                    'range': ['2018-01-01', '2018-01-31']
                }
            }

        },
        'msg': {
            'name': {
                # 默认消息提示，如果找不到下一级的消息提示则提示该消息
                '_default': '默认消息提示',

                # 覆盖默认提示
                'enum': '姓名必须在huen和rean中'
            },
            'confirm_password': {
                'equal_to': '两次输入的密码不一样。'
            },
            'birth': {
                'date': '生日输入不正确.'
            }
        }
    }

    obj = {
        'name': 'rean',
        'password': '674038364@qq.com',
        'home_page': 'http://baidu.com',
        'confirm_password': '674038364@qq.com',
        'score': 200,
        'address': '花22dasdsadsadsadasdsad道',
        'address1': '22',
        'ipaddr': '255.255.255.255',
        'birth': '2018-01-43'
    }
    print(check.validate(rule, obj))
    print(check.id_card('52130199310062011'))


validate/check.py
# coding=utf8

from validate.utils import id_card, enum, url, email, equal_to, min_value, max_value, range_value, min_length, \
    max_length, range_length, pattern, custom, ipv4, ipv6, date, tel, post_code

__all__ = ['default_msg',
           'validate']

default_msg = {
    'required': u'字段必填.',
    'enum': u'字段必须在给定的范围内.',
    'type': u'字段类型不正确.',
    'email': u'字段不符合电子邮件格式.',
    'url': u'字段不是合法的url.',
    'equal_to': u'字段两次输入的值不相等.',
    'min': u'字段的值太小.',
    'max': u'字段的值太大.',
    'range': u'字段的值不在范围内.',
    'min_length': u'字段太短.',
    'max_length': u'字段太长.',
    'range_length': u'字段的长度不在范围内.',
    'pattern:': u'字段符合正则表达式.',
    'custom': u'字段验证不通过.',
    'id_card': u'字段不是合法的身份证号.',
    'ipv4': u'字段不是一个合法的ip4地址.',
    'ipv6': u'字段不是一个合法的ip6地址.',
    'date': u'字段不是一个合法的时间格式.',
    'tel': u'字段不是一个合法的电话号码',
    'post_code': u'字段不是合法的邮编'
}


class ERRMSG(dict):
    """
    获取错误提示
    查找顺序:
    1.自定义每个属性的提示
    2.字段默认提示（_default）
    3.全局默认提示
    """

    def __init__(self, msg, **kwargs):
        super(ERRMSG, self).__init__(**kwargs)
        self.msg = msg

    def __getitem__(self, item):
        field, attr = item.split('.')
        if field in self.msg:
            field = self.msg[field]
            if attr in field:
                return field[attr]
            else:
                if '_default' in field:
                    return field['_default']
                else:
                    pass
        else:
            pass

        try:
            return default_msg[attr]
        except KeyError:
            raise Exception('没有设置默认提示.')


def validate(rules, obj):
    """
    验证函数
    :param rules:验证规则
    :param obj:验证对象
    :param args:错误提示
    :return:
    """
    validate_all = False
    if 'all_validate' in rules and rules['all_validate'] is True:
        validate_all = True
    if 'fields' not in rules or type(rules['fields']) != dict:
        raise Exception('字段参数异常')
    fields = rules['fields']

    if 'msg' in rules:
        errmsg = ERRMSG(rules['msg'])
    else:
        errmsg = ERRMSG({})

    error_list = []
    pass_list = []
    valid_list = list(fields.keys())

    while len(valid_list) != 0:
        if validate_all is False and len(error_list) > 0:
            return True if error_list.__len__() == 0 else False, error_list, pass_list

        field = valid_list[0]
        result, attr = __val_field(field, fields[field], obj)
        if result:
            pass_list.append(field)
        else:
            error_list.append((field, errmsg['.'.join([field, attr])]))
        del valid_list[0]

    return True if error_list.__len__() == 0 else False, error_list, pass_list


def __val_field(field, rule, obj):
    """
    验证每个字段
    :param field:字段名
    :param rule:验证规则
    :param obj:待验证对象
    :return:
    """
    if 'required' in rule:
        if rule['required'] and field not in obj:
            return False, 'required'

    try:
        value = obj[field]
    except KeyError:
        return False, 'required'

    if 'enum' in rule:
        if enum(value, rule['enum']) is False:
            return False, 'enum'

    if 'type' in rule:
        if rule['type'] != type(value) is False:
            return False, 'type'

    if 'email' in rule and rule['email'] is True:
        if email(value) is False:
            return False, 'email'

    if 'url' in rule and rule['url'] is True:
        if url(value) is False:
            return False, 'url'

    if 'equal_to' in rule:
        if equal_to(obj[rule['equal_to']], value) is False:
            return False, 'equal_to'

    if 'min' in rule:
        if min_value(value, rule['min']) is False:
            return False, 'min'

    if 'max' in rule:
        if max_value(value, rule['max']) is False:
            return False, 'max'

    if 'range' in rule:
        if range_value(value, rule['range']) is False:
            return False, 'range'

    if 'min_length' in rule:
        if min_length(value, rule['min_length']) is False:
            return False, 'min_length'

    if 'max_length' in rule:
        if max_length(value, rule['min_length']) is False:
            return False, 'max_length'

    if 'range_length' in rule:
        if range_length(value, rule['range_length']) is False:
            return False, 'range_length'

    if 'pattern' in rule:
        if pattern(rule['pattern'], value) is None:
            return False, 'pattern'

    if 'custom' in rule:
        if custom(field, value, rule['custom']) is False:
            return False, 'custom'

    if 'id_card' in rule and rule['id_card'] is True:
        if id_card(obj['field']) is False:
            return False, 'id_card'

    if 'ipv4' in rule and rule['ipv4'] is True:
        if not ipv4(value):
            return False, 'ipv4'

    if 'ipv6' in rule and rule['ipv6'] is True:
        if not ipv6(value):
            return False, 'ipv6'

    if 'date' in rule:
        date_rule = rule['date']
        if not date(value, date_rule):
            return False, 'date'

    if 'tel' in rule and rule['tel'] is True:
        if not tel(value):
            return False, 'tel'

    if 'post_code' in rule and rule['post_code'] is True:
        if not post_code(value):
            return False, 'post_code'

    return True, field


if __name__ == '__main__':
    print(id_card('5555555'))
    print(id_card('522130931006201'))
    print(id_card('522130199310062011'))
    print(ipv4('192.168.168.1'))
    print(ipv6('fe80::356c:4a6f:64f4:c88e'))


validate/utils.py
# coding=utf8

import re

try:
    from id_card import id_card
    from vdate import date
except ImportError:
    from .id_card import id_card
    from .vdate import date

__all__ = [
    'id_card',
    'date',
    'email',
    'enum',
    'url',
    'equal_to',
    'min_value',
    'max_value',
    'range_value',
    'min_length',
    'max_length',
    'range_length',
    'pattern',
    'custom',
    'ipv4',
    'ipv6',
    'tel',
    'post_code'
]


def email(val):
    """
    验证电子邮件
    :param val:电子邮件地址
    :return:
    """
    return pattern(val, '^[A-Za-z\d]+([-_.][A-Za-z\d]+)*@([A-Za-z\d]+[-.])+[A-Za-z\d]+$')


def enum(val, coll):
    """
    验证枚举类型
    :param val:
    :param coll:
    :return:
    """
    return val in coll


def url(val):
    """
    验证URL
    :param val:
    :return:
    """
    return pattern(val, '(https?|ftp|file)://[-A-Za-z0-9+&@#/%?=~_|!:,.;]+[-A-Za-z0-9+&@#/%=~_|]')


def equal_to(left, right):
    """
    验证两个字段是否相等
    :param left:
    :param rigth:
    :return:
    """
    return left == right


def min_value(val, limit):
    """
    验证是否小于最小值，先转换为数值比较，如果转换不成功，则转换为字符串比较
    :param val:
    :return:
    """
    try:
        a, b = int(val), int(limit)
    except ValueError:
        a, b = str(val), str(limit)
    return a > b


def max_value(val, limit):
    """
    验证是否大于最大值，先转换为数值比较，如果转换不成功，则转换为字符串比较
    :param val:
    :param limit:
    :return:
    """
    try:
        a, b = int(val), int(limit)
    except ValueError:
        a, b = str(val), str(limit)
    return a < b


def range_value(val, limit):
    """
    验证给定值是否在范围内,先转换为数值比较，如果转换不成功，则转换为字符串比较
    :param val:
    :param limit:
    :return:
    """
    if (isinstance(limit, tuple) or isinstance(limit, list)) and len(limit) == 2:
        try:
            a, b, c = int(limit[0]), int(val), int(limit[1])
        except ValueError:
            a, b, c = str(limit[0]), str(val), str(limit[1])

        return a <= b <= c

    return False


def min_length(val, limit):
    """
    验证长度是否小于给定值
    :param val:
    :param length:
    :return:
    """
    return str(val).__len__() >= limit


def max_length(val, limit):
    """
    验证长度是否大于给定值
    :param val:
    :param limit:
    :return:
    """
    return str(val).__len__() <= limit


def range_length(val, limit):
    """
    验证长度是否在范围内
    :param val:
    :param limit:
    :return:
    """
    if (isinstance(limit, tuple) or isinstance(limit, list)) and len(limit) == 2:
        return limit[0] <= len(str(val)) <= limit[1]

    return False


def pattern(val, limit):
    """
    正则验证值是否在范围内
    :param val:
    :param limit:
    :return:
    """
    result = re.match(limit, val)
    return True if result is not None else False


def custom(field, val, limit):
    """
    使用自定义函数验证
    :param field: 字段名
    :param val:值
    :param limit: 自定义函数
    :return:
    """
    if callable(limit) is True:
        return limit(field, val)
    return False


def tel(val):
    """
    验证中国手机号,+86格式的不验证，指验证11为手机号:
    :param val:要验证的电话号码
    :return:
    """
    return pattern(str(val), '^1[34578]\d{9}$')


def post_code(val):
    """
    验证中国邮政编码
    :param val:
    :return:
    """
    return pattern(str(val), '^[1-9]\d{5}$')


def ipv4(val):
    """
    验证IPV4
    :param val:
    :return:
    """
    return pattern(val,
                   '^(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)\.(25[0-5]|2[0-4]\d|[01]?\d\d?)$')


def ipv6(val):
    """
    验证IPV6
    :param val:
    :return:
    """
    return pattern(val,
                   '^((([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){6}:[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){5}:([0-9A-Fa-f]{1,4}:)?[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){4}:([0-9A-Fa-f]{1,4}:){0,2}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){3}:([0-9A-Fa-f]{1,4}:){0,3}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){2}:([0-9A-Fa-f]{1,4}:){0,4}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){6}((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|(([0-9A-Fa-f]{1,4}:){0,5}:((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|(::([0-9A-Fa-f]{1,4}:){0,5}((\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b)\.){3}(\b((25[0-5])|(1\d{2})|(2[0-4]\d)|(\d{1,2}))\b))|([0-9A-Fa-f]{1,4}::([0-9A-Fa-f]{1,4}:){0,5}[0-9A-Fa-f]{1,4})|(::([0-9A-Fa-f]{1,4}:){0,6}[0-9A-Fa-f]{1,4})|(([0-9A-Fa-f]{1,4}:){1,7}:))$')


id_card('522130199310062011')
if __name__ == '__main__':
    print(id_card('522130199310062011'))
    print(email('674038364@qq.com'))
    print(range_value(2, [1, 3]))
    print(min_length(22, 1))
    print(ipv4('192.168.168.1'))
    print(ipv6('fe80::356c:4a6f:64f4:c88e'))


validate/vdate.py
# coding=utf8

import time

__all__ = ['date']


def date(val, rule):
    """
    验证日期,格式验证，大小验证,大小验证和范围验证转换为时间戳验证
    :param val:
    :param rule:
    :return:
    """
    err = '时间格式不正确.'

    if 'format' not in rule:
        raise Exception('没有设置时间验证的格式.')
    try:
        fat = rule['format']
        time_str = time.strptime(val, fat)
    except ValueError:
        return False

    day = time.mktime(time.strptime(time.strftime(fat), fat))  # 根据格式格式化当前时间
    val = time.mktime(time_str)
    if 'min' in rule:
        if rule['min'] != 1:
            try:
                day = time.mktime(time.strptime(rule['min'], fat))
            except ValueError:
                raise Exception(err)
        if val < day:
            return False

    if 'max' in rule:
        if rule['max'] != 1:
            try:
                day = time.mktime(time.strptime(rule['max'], fat))
            except ValueError:
                raise Exception(err)
        if val > day:
            return False

    if 'range' in rule:
        limit = rule['range']
        try:
            a, c = time.mktime(time.strptime(limit[0], fat)), time.mktime(time.strptime(limit[1], fat))
            if a > val or c < val:
                return False
        except ValueError:
            raise ValueError(err)

    return True


if __name__ == '__main__':
    birth = {
        'format': '%Y-%m-%d',
        'min': 1,  # 1，最小值是当前时间，其他字符串
        'max': '2018-01-31',  # 1，最大值是当前时间，其他字符串
        'range': ['2018-01-01', '2018-01-31']
    }
    print(date('2018-01-04', birth))


validate/id_card.py
# coding=utf8

import time
import re

__all__ = ['id_card']


def id_card(val):
    """
    1.验证长度
    2.验证省份
    3.验证年龄
    5.验证校验位
    :param val:
    :return:
    """
    return check_length(val) and check_province(val) and check_birthday(val) and check_last(val)


def check_length(val):
    return True if re.match('(^\d{15}$)|(^\d{17}(\d|X)$)', val) is not None else False


def check_province(val):
    """
    验证身份证的省份
    :param val:
    :return:
    """
    city = {
        '11': '北京',
        '12': '天津',
        '13': '河北',
        '14': '山西',
        '15': '内蒙古',
        '21': '辽宁',
        '22': '吉林',
        '23': '黑龙江',
        '31': '上海',
        '32': '江苏',
        '33': '浙江',
        '34': '安徽',
        '35': '福建',
        '36': '江西',
        '37': '山东',
        '41': '河南',
        '42': '湖北',
        '43': '湖南',
        '44': '广东',
        '45': '广西',
        '46': '海南',
        '50': '重庆',
        '51': '四川',
        '52': '贵州',
        '53': '云南',
        '54': '西藏',
        '61': '陕西',
        '62': '甘肃',
        '63': '青海',
        '64': '宁夏',
        '65': '新疆',
        '71': '台湾',
        '81': '香港',
        '82': '澳门',
        '91': '国外'
    }
    return val[0:2] in city


def verify_birthday(year):
    """
    验证身份证的年龄正不正确
    :param year:
    :return:
    """
    t = time.localtime()
    now_year = t.tm_year
    age = now_year - int(year)
    return 0 <= age < 130


def check_birthday(val):
    """
    验证生日是否
    :param val:
    :return:
    """
    length = len(val)
    if length == 15:
        # 身份证15位时，次序为省（3位）市（3位）年（2位）月（2位）日（2位）校验位（3位），皆为数字
        year = '19' + val[6:8]
        return verify_birthday(year)
    if length == 18:
        # 身份证18位时，次序为省（3位）市（3位）年（4位）月（2 位）日（2位）校验位（4位），校验位末尾可能为X
        return verify_birthday(val[6:10])
    return False


def check_last(val):
    """
    验证最后一个校验位
    :param val:
    :return:
    """
    if len(val) == 15:
        val = fifteen_to_eighteen(val)
    arr_int = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
    arr_ch = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2']
    card_temp = 0
    for i in range(0, 17):
        card_temp += int(val[i]) * arr_int[i]
    val_num = arr_ch[card_temp % 11]
    return val[17] == val_num


def fifteen_to_eighteen(val):
    """
    将15位身份证转为18位身份证
    :param val:
    :return:
    """
    if len(val) == 15:
        arr_int = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2]
        arr_ch = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2']
        card_temp = 0
        val = val[0:6] + '19' + val[6:]
        for i in range(0, 17):
            card_temp += int(val[i]) * arr_int[i]
        val += arr_ch[card_temp % 11]
    return val


if __name__ == '__main__':
    print(id_card('5555555'))
    print(id_card('522130931006201'))
    print(id_card('522130199310062011'))


example.py
# coding=utf8

from validate import check


def custom(filed, value):
    """
    自定义验证函数
    :param filed:字段名
    :param value:字段值
    :return:
    """
    if filed == 'ipaddr':
        return check.pattern(value, '^(25[0-5]|2[0-4]\d|[0-1]?\d?\d)(\.(25[0-5]|2[0-4]\d|[0-1]?\d?\d)){3}$')


if __name__ == '__main__':
    # 覆盖默认提示
    check.default_msg['required'] = u'字段必填'
    # 提示查找顺序
    # 1.自定义每个属性的提示
    # 2.字段默认提示（_default）
    # 3.全局默认提示

    rule = {
        'all_validate': True,  # 全部验证在返回，否则只要有错就返回
        'fields': {
            'name': {
                'required': True,
                'enum': ['rean', 'huen'],
                'type': str,
            },
            'email': {
                'required': True,
                'email': True
            },
            'home_page': {
                'url': True
            },

            'password': {
                'range_length': [16, 20]
            },
            'confirm_password': {
                'equal_to': 'password',  # password 字段一样
            },
            'age': {
                'min': 0,
                'max': 130
            },
            'score': {
                'range': [0, 100]
            },
            'address': {
                'min_length': 4,
                'max_length': 40,
                'range_length': (4, 40)
            },
            'ipaddr': {
                'custom': custom  # 自定义函数验证
            },
            'phone_number': {
                'tel': True
            },
            'post_code': {
                'post_code': True
            },
            'birth': {
                'date': {
                    'format': '%Y-%m-%d',
                    'min': 1,  # 1，最小值是当前时间，其他字符串
                    'max': '2018-01-31',  # 1，最大值是当前时间，其他字符串
                    'range': ['2018-01-01', '2018-01-31']
                }
            }
        },
        'msg': {
            'name': {
                # 默认消息提示，如果找不到下一级的消息提示则提示该消息
                '_default': u'默认消息提示',
                # 覆盖默认提示
                'enum': u'姓名必须在huen和rean中'
            },
            'confirm_password': {
                'equal_to': u'两次输入的密码不一样。'
            },
            'birth': {
                'date': u'生日输入不正确.'
            }
        }
    }

    value = {
        'name': 'rean',
        'email': 'rean.ren@qq.com',
        'password': '1234567890123456',
        'home_page': 'http://baidu.com',
        'confirm_password': '1234567890123456',
        'age': 30,
        'score': 90,
        'address': '花22dasdsadsadsadasdsad道',
        'ipaddr': '255.255.255.255',
        'phone_number': '13984927617',
        'post_code': '564500',
        'birth': '2018-01-43'
    }
    print(check.validate(rule, value))
    print(check.id_card('52130199310062011'))

    # (False, [('address', '字段太长.'), ('birth', '生日输入不正确.')], ['name', 'email', 'home_page', 'password', 'confirm_password', 'age', 'score', 'ipaddr', 'phone_number', 'post_code'])
    # False
